### ISSUE TITLE
Failures attempting to build async_kernel with Ocaml 4.07

### ISSUE THREAD

**Issue by enolan** [timbertson/opam2nix-packages#48]
I see a permissions failure due to sandboxing and a different failure with sandboxing turned off. The command I'm running is: `nix-build -E 'with (import <nixpkgs> {}); let opam2nix = pkgs.callPackage ./default.nix {}; in opam2nix.buildPackage "async_kernel" {ocaml = pkgs.ocaml-ng.ocamlPackages_4_07.ocaml;}'`, where `default.nix` is `nix/release/default.nix` from this repo.

With sandboxing on I see:
```
building '/nix/store/n35brj24ajinfhf9whd9lx94p5jil7yg-ocaml-4.07.1.drv'...
error: moving build output '/nix/store/p2rw46xaf6xr95sgzphka09bl460s37g-ocaml-4.07.1' from the sandbox to the Nix store: Permission denied
```

When it tries to build `/nix/store/n35brj24ajinfhf9whd9lx94p5jil7yg-ocaml-4.07.1.drv`.

With it off the compiler builds fine but it fails later on when trying to build `/nix/store/v5qd3ag27rj8fnqpv0b90ji3nvpmp6zb-ocaml-migrate-parsetree-1.2.0.drv`:

```
unpacking sources
unpacking source archive /nix/store/mb05szkq5xn75vz6gnch7973ng8smr4q-ocaml-migrate-parsetree-v1.2.0.tbz
source root is ocaml-migrate-parsetree-v1.2.0
setting SOURCE_DATE_EPOCH to timestamp 1545211134 of file ocaml-migrate-parsetree-v1.2.0/tools/pp_rewrite.mll
patching sources
configuring
building
+ /nix/store/ingw9z6fvnaa42pzx91hyh35cr0y99cq-opam2nix-0.5.0/bin/opam2nix invoke build
Loading /nix/store/xjsw88pfrph5gk8757xg2clp6xw64ghr-opam
creating /nix/store/m662nmisk08ygygggyk04rv96wyd4dck-ocaml-migrate-parsetree-1.2.0
creating /nix/store/m662nmisk08ygygggyk04rv96wyd4dck-ocaml-migrate-parsetree-1.2.0/bin
creating /nix/store/m662nmisk08ygygggyk04rv96wyd4dck-ocaml-migrate-parsetree-1.2.0/lib/stublibs
+ dune build -p ocaml-migrate-parsetree -j 1
File "src/dune", line 5, characters 40-52:
5 |  (libraries compiler-libs.common result ppx_derivers)
                                            ^^^^^^^^^^^^
Error: Library "ppx_derivers" not found.
Hint: try: dune external-lib-deps --missing -j 1 -p ocaml-migrate-parsetree @@default
Command failed.
builder for '/nix/store/v5qd3ag27rj8fnqpv0b90ji3nvpmp6zb-ocaml-migrate-parsetree-1.2.0.drv' failed with exit code 1
```

Looking at the `.drv` it does seem to include `ppx_derivers` as a dependency.

I'm running Ubuntu and using the `nixpkgs-unstable` channel.

**Comment by timbertson** [comment]
Ahh, I think I've found the bug. There are two ways that we inject the requested `ocaml` version (which we then extend with support for OCAMLPATH and OCAML_TOPLEVEL_PATH). The orginal method used an attribute, which allows us to look up the derivation at build time and use `overrideAttributes` on it. The new method supplies a path to whatever `.drv` you like, which means it can be an arbitrary derivation, not necessarily an attribute in nixpkgs. The downside is you can't `overrideAttributes` on an imported `.drv`, so it has to wrap the derivation instead.

Unfortunately, it looks like `propagatedBuildInputs` isn't working for the latter mode, which is the thing that provides OCAMLPATH setup (thus not finding dependencies).

As a workaround, you can replace:

`ocaml = pkgs.ocaml-ng.ocamlPackages_4_07.ocaml;`

...with:

`ocamlAttr = "ocaml-ng.ocamlPackages_4_07.ocaml";`

Until I get to the bottom of why this is broken...

**Comment by timbertson** [comment]
OK, both issues fixed in 8a36bf26 (with release/ updated in the following commit). Thanks for reporting!

**Comment by enolan** [comment]
Thanks, Tim! Working now.
